src/Xrefcheck/Verify.hs

```diff
@@ -426,94 +448,104 @@ verifyReference
-        fixable = any isFixable ves
-
-    extractRetryAfterInfo :: VerifyResult VerifyError -> Maybe RetryAfter
-    extractRetryAfterInfo = \case
-      VerifyResult [ExternalHttpTooManyRequests retryAfter] -> retryAfter
+      case res of
+        -- Success
+        Right () -> do
+          let moveProgressOnSuccess = composeFuncList
+                [ incProgress
+                , whenFunc (errorsOccured rc) decProgressFixableErrors
+                ]
+          modifyProgressRef now Nothing moveProgressOnSuccess
+          pure res
+        Left err -> do
+          setOfReturned429 <- addDomainIf429 domainsReturned429Ref err
+          case extractRetryInfo setOfReturned429 rc err of
+            -- Unfixable
+            Nothing -> do
+              let moveProgressOnUnfixable = composeFuncList
+                    [ incProgress
+                    , if errorsOccured rc
+                      then fixableToUnfixable
+                      else incProgressUnfixableErrors
+                    ]
+              modifyProgressRef now Nothing moveProgressOnUnfixable
+              pure res
+            -- Fixable, retry
+            Just (mbCurrentRetryAfter, counterModifier) -> do
+              let toSeconds = \case
+                    Seconds s -> s
+                    -- Calculates the seconds left until @Retry-After@ date.
+                    -- Defaults to 0 if the date has already passed.
+                    Date date | utcTimeToTimeSecond date >= now -> utcTimeToTimeSecond date -:- now
+                    _ -> sec 0
+
+              let currentRetryAfter = fromMaybe (ncDefaultRetryAfter cNetworking) $
+                    fmap toSeconds mbCurrentRetryAfter
+
+              let moveProgressOnFixable = whenFunc (rcTotalRetries == 0) incProgressFixableErrors
+              modifyProgressRef now (Just currentRetryAfter) moveProgressOnFixable
+              threadDelay currentRetryAfter
+              retryVerification
+                (counterModifier rc)
+                resIO
+
+    modifyProgressRef :: Time Second -> Maybe (Time Second) -> (Progress Int -> Progress Int) -> IO ()
+    modifyProgressRef now mbRetryAfter moveProgress = atomicModifyIORef' progressRef $ \VerifyProgress{..} ->
+      ( if isExternal $ locationType rLink
+        then VerifyProgress{ vrExternal =
+          let vrExternalAdvanced = moveProgress vrExternal
+          in case mbRetryAfter of
+             Just retryAfter -> case pTaskTimestamp vrExternal of
+                    Just (TaskTimestamp ttc start)
+                      | retryAfter +:+ now <= ttc +:+ start -> vrExternalAdvanced
+                    _ -> setTaskTimestamp retryAfter now vrExternalAdvanced
+             Nothing -> vrExternalAdvanced, .. }
+        else VerifyProgress{ vrLocal = moveProgress vrLocal, .. }
+      , ()
+      )
+
+    addDomainIf429 :: IORef (S.Set Text) -> VerifyError -> IO (S.Set Text)
+    addDomainIf429 setRef err = atomicModifyIORef' setRef $ \s ->
```

Comment:
I'm not sure how much sense this concern has, but maybe it will be interesting to investigate.

The potential issue with `atomicModifyIORef'` is that it may cause high contention. If we get 1000 simultaneous responses with 429 and invoke this function repeatedly, it may end up with different calls hindering each other, because AFAIU `atomicModifyIORef` is implemented as compare-and-set, and each thread will get value, perform `S.insert` computation, only to find out that during that time another thread tried to insert something and thus the result is invalidated and computation must be repeated. If this is true, we may get quadratic execution time.

At least, I have been witnessing such issues with STM once.

Could you set up a simple experiment with some minimalistic example using `atomicModifyIORef'` and try?

Reply:
#251 has been created for this, we will re-consider things there.